synapse/storage/data_stores/main/events.py

```diff
@@ -176,6 +218,11 @@ def _persist_events_and_state_updates(
             )
             persist_event_counter.inc(len(events_and_contexts))
 
+            for event, _ in events_and_contexts:
+                self.store.get_unread_message_count_for_user.invalidate_many(
+                    (event.room_id,),
+                )
+
```

Comment:
does this not also need to happen on worker instances? think you need to hook into `invalidate_caches_for_event`.

Also... given we already have a loop over `events_and_contexts` five lines down, any reason not to put this in there?

also also... can we skip it if `backfilled` ?

Reply:
> Also... given we already have a loop over events_and_contexts five lines down, any reason not to put this in there?

I don't remember exactly, but I assume I just missed that loop.

> also also... can we skip it if backfilled ?

I don't think we should, because we might need to count backfilled events as unread. FWIW I'm usually trying to be consistent with how we handle push actions since it's a very similar feature, and I can see that we invalidate `get_unread_event_push_actions_by_room_for_user` regardless of `backfilled`.